src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java

```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
     return taskIndex.getTask(taskId).getTestingCopyBuilderInfo();
   }
 
+  @Override
+  public List<String> listFiles(String taskId, Boolean usingTestingVersion, String variant)
+      throws TaskNotFoundException, TaskMissingVariantException, TaskStorageException {
+    Task t = taskIndex.getTask(taskId);
+    LOG.info("List files " + t + ", " + usingTestingVersion + ", " + variant);
+    try (TaskCopy c = usingTestingVersion ? t.acquireTestingCopy() : t.acquireRegisteredCopy()) {
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      return c.listSkeleton(variant);
+    }
+  }
+
+  @Override
+  public Response readFile(String taskId, String fileName, Boolean usingTestingVersion,
+                           String variant)
+      throws TaskNotFoundException, TaskMissingVariantException {
+    Task t = taskIndex.getTask(taskId);
+    try (TaskCopy c = usingTestingVersion ? t.acquireTestingCopy() : t.acquireRegisteredCopy()) {
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      File filePath = new File(c.getSkeletonLocation(variant), fileName);
+      return Response.ok(filePath, MediaType.APPLICATION_OCTET_STREAM).build();
```

Comment:
This just returns the file name doesn't it - don't you need something like RepoController.readFile?

Reply:
It seems Response actually knows to return the file contents when given a File. I've added a comment to that effect.

Reply:
ah. that's clever - thanks for the comment - that will help future-me